xray/services/buildscan.go

```diff
@@ -0,0 +1,146 @@
+package services
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/jfrog/jfrog-client-go/artifactory/services/utils"
+	"github.com/jfrog/jfrog-client-go/auth"
+	"github.com/jfrog/jfrog-client-go/http/jfroghttpclient"
+	clientutils "github.com/jfrog/jfrog-client-go/utils"
+	"github.com/jfrog/jfrog-client-go/utils/errorutils"
+	"github.com/jfrog/jfrog-client-go/utils/log"
+)
+
+const (
+	buildScanAPI                        = "api/v2/ci/build"
+	xrayScanBuildNotSelectedForIndexing = "is not selected for indexing"
+	XrayScanBuildNoFailBuildPolicy      = "No Xray “Fail build in case of a violation” policy rule has been defined on this build"
+	projectKeyQueryParam                = "projectKey="
+)
+
+type BuildScanService struct {
+	client      *jfroghttpclient.JfrogHttpClient
+	XrayDetails auth.ServiceDetails
+}
+
+// NewBuildScanService creates a new service to scan Binaries and VCS projects.
+func NewBuildScanService(client *jfroghttpclient.JfrogHttpClient) *BuildScanService {
+	return &BuildScanService{client: client}
+}
+
+func (bs *BuildScanService) BuildScan(params XrayBuildParams) (string, error) {
+	httpClientsDetails := bs.XrayDetails.CreateHttpClientDetails()
+	utils.SetContentType("application/json", &httpClientsDetails.Headers)
+	requestBody, err := json.Marshal(params)
+	if err != nil {
+		return "", errorutils.CheckError(err)
+	}
+	url := bs.XrayDetails.GetUrl() + buildScanAPI
+
+	resp, body, err := bs.client.SendPost(url, requestBody, &httpClientsDetails)
+	if err != nil {
+		return "", err
+	}
+
+	if err = errorutils.CheckResponseStatus(resp, http.StatusOK, http.StatusCreated); err != nil {
+		return "", errorutils.CheckError(errorutils.GenerateResponseError(resp.Status, clientutils.IndentJson(body)))
+	}
+	buildScanResponse := RequestBuildScanResponse{}
+	if err = json.Unmarshal(body, &buildScanResponse); err != nil {
+		return "", errorutils.CheckError(err)
+	}
+	buildScanInfo := buildScanResponse.Info
+	if strings.Contains(buildScanInfo, xrayScanBuildNotSelectedForIndexing) {
+		return "", errorutils.CheckErrorf(buildScanResponse.Info)
+	}
+	log.Info(buildScanInfo)
+	return buildScanInfo, nil
+}
+
+func (bs *BuildScanService) GetBuildScanResults(params XrayBuildParams) (*BuildScanResponse, error) {
+	requestUrl := fmt.Sprintf("%s%s/%s/%s", bs.XrayDetails.GetUrl(), buildScanAPI, params.BuildName, params.BuildNumber)
+	if params.Project != "" {
+		requestUrl += projectKeyQueryParam + params.Project
+	}
+	syncMessage := fmt.Sprintf("Sync: Get Build Scan results. Build:%s/%s...", params.BuildName, params.BuildNumber)
+	httpClientsDetails := bs.XrayDetails.CreateHttpClientDetails()
+	utils.SetContentType("application/json", &httpClientsDetails.Headers)
+
+	//The scan request may take some time to complete. We expect to receive a 202 response, until the completion.
+	ticker := time.NewTicker(defaultSyncSleepInterval)
```

Comment:
It looks to me that this waiting mechanis, with the ticker is more compex that it should be.
Why don't we just have a loop with a fixed number of iterations (which is the wait timeout actually) and a 5 seconds sleep inside it. We also don't need to spawn a new go routine for it.